### PR TITLE
Create a lossy index of resource -> ClassPathElement mapping

### PR DESCRIPTION
**Pull request by gsmet**
> [!NOTE]
> This PR is part of a coordinated effort to address the OOM reported in https://github.com/quarkusio/quarkus/issues/42471:
> - https://github.com/quarkusio/quarkus/pull/42492
> - https://github.com/quarkusio/quarkus/pull/42525
> - https://github.com/quarkusio/quarkus/pull/42560

The size of the `ClassLoaderState` of the `QuarkusClassLoader` can be extremely problematic (I have seen it regularly at 10+ MB in the various heap dumps I studied these past months).

I made a first attempt at fixing it a few months ago but wasn't really happy with the approach. I approached the problem differently here.

The idea of this patch is to have a lossy index instead and try to find a good compromise between speed and memory usage.
And that's the only compromise we make as each `ClassPathElement` has a comprehensive index of what it contains so we can narrow down the candidates using the lossy index and then fully check the few ``ClassPathElement``s candidates.

I personally think it makes things a bit more readable too but YMMV.

This is related to issue https://github.com/quarkusio/quarkus/issues/42471 .

In this particular example:
- the old `state` is 13 MB+
- the new index is less than 3 MB

So we end up with 10 less MB per `QuarkusClassLoader` in this case (obviously this will depend on the number of classes in your app).

As for speed, I made a few tests and things are not noticeably slower when starting an application in dev mode from what I could see. I made the implementation of `ClassPathResourceIndex#getResourceKey()` faster in a follow up commit.

One thing that I noticed (but it's preexisting) is that I saw this call appear a few times in the flamegraph:
https://github.com/quarkusio/quarkus/blob/main/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/PathTreeClassPathElement.java#L105-L107

When calling this from the CL we could bypass this check as we know the resource is there. But I'm not sure how to make it evolve from an API perspective. /cc @aloubyansky 

This is better reviewed commit per commit as I added a few more things.

Situation before this patch:

Screenshot from 2024-08-16 14-17-17

Situation after this patch:

Screenshot from 2024-08-16 14-15-55

### CODE REVIEW

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java [hunk 1]
```diff
@@ -183,10 +183,12 @@ private void addCpElement(QuarkusClassLoader.Builder builder, ResolvedDependency
             //we always load this from the parent if it is available, as this acts as a bridge between the running
             //app and the dev mode code
             builder.addParentFirstElement(element);
+            builder.addNormalPriorityElement(element);
         } else if (dep.isFlagSet(DependencyFlags.CLASSLOADER_LESSER_PRIORITY)) {
             builder.addLesserPriorityElement(element);
+        } else {
+            builder.addNormalPriorityElement(element);
         }
-        builder.addElement(element);
```
Comment (gsmet) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/CuratedApplication.java:189]: The semantic have changed a bit here: we used to store all the elements in `elements` and then do some logic to figure out if they were in `lesserPriorityElements` when building the index.
I chose to simplify things and have a clearer semantic.

Thus why the logic has changed here and also why I renamed the method.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+ * of having a lossy index referencing the start of the resource path to reduce the number of ClassPathElements to go through.
+ * This index still returns exact references as we check the ClassPathElements themselves for the presence of the resource.
+ * <p>
+ * In the particular example above, we went down to less than 3 MB for the index, storing ~600 entries instead of 86000+ in the
+ * mapping.
+ * <p>
+ * We try to be clever as to how we build the resource key to reduce the number of misses. It might need further tuning in the
+ * future.
+ */
+public class ClassPathResourceIndex {
```
Comment (gsmet) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java:29]: The removal of `ClassLoaderState` and the addition of this class is the whole purpose of this patch. The rest is really to be able to make this happen.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java [hunk 3]
```diff
@@ -96,7 +95,11 @@ public static boolean isResourcePresentAtRuntime(String resourcePath) {
     }
 
     private final String name;
-    private final List<ClassPathElement> elements;
+    // the ClassPathElements to consider are normalPriorityElements + lesserPriorityElements
+    private final List<ClassPathElement> normalPriorityElements;
+    private final List<ClassPathElement> lesserPriorityElements;
```
Comment (gsmet) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:100]: As explained in https://github.com/quarkusio/quarkus/pull/42525/files#r1719580032, I changed the semantic a bit here.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java [hunk 4]
```diff
@@ -588,27 +602,30 @@ public List<ClassPathElement> getElementsWithResource(String name) {
...
+
+        List<ClassPathElement> classPathElements = getClassPathResourceIndex().getClassPathElements(name);
+        ret.addAll(classPathElements);
+
+        if (!parentFirst && !localOnly && parent instanceof QuarkusClassLoader) {
+            ret.addAll(((QuarkusClassLoader) parent).getElementsWithResource(name));
         }
-        ret.addAll(Arrays.asList(classPathElements));
+
         return ret;
```
Comment (gsmet) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:620]: The behavior of this method was extremely odd as it was always returning the parent resources first. I fixed it to honor parent first in a subsequent commit (hoping it won't break anything - the rest have already been thoroughly tested on CI).

**Comment by geoand** [comment]
I'll have a good look at this when I'm back (in 10 days or so)

**Comment by gsmet** [comment]
This is ready for review, it passed CI on my fork a few times.

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java [hunk 5]
```diff
@@ -80,6 +80,13 @@ static PathTree ofArchive(Path archive, PathFilter filter) {
         return ArchivePathTree.forPath(archive, filter);
     }
 
+    /**
+     * Whether the PathTree provides local resources.
+     * <p>
+     * For instance a directory or a file provides local resources. A jar does not.
+     */
```
Comment (geoand) [inline independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java:87]: I don't really understand what local resources are

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java (hunk 2 above)
Comment (geoand) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java:29]: It might just be the fact that it's my first day back from PTO, but just by reading the Javadoc, I don't understand how this index is supposed to work. I do understand what problem it tries to solve vs. previous approach, but I don't understand how it does it.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    private static String getResourceKey(String resource) {
+        // we don't really care about this part, it can be slower
+        if (resource.startsWith(META_INF_MAVEN)) {
+            return META_INF_MAVEN;
+        }
+        if (resource.startsWith(META_INF_SERVICES)) {
+            // for services, we want to reference the full path
+            return resource;
+        }
+        StringBuilder prefix = new StringBuilder();
```
Comment (geoand) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java]: Any chance we can make an okay guess of the length?

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java [hunk 7]
```diff
@@ -355,46 +333,61 @@ public URL getResource(String unsanitisedName) {
...
+            if (url != null) {
+                return url;
+            }
+        }
+        return parent.getResource(unsanitisedName);
+    }
+
+    private static URL getClassPathElementResourceUrl(List<ClassPathElement> classPathElements, String name,
+            boolean endsWithTrailingSlash) {
+        for (ClassPathElement classPathElement : classPathElements) {
```
Comment (geoand) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:377]: Can we use a good old `for(i=0...)` here in order to cut down on allocations?

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java [hunk 8]
```diff
@@ -592,23 +600,20 @@ public List<ClassPathElement> getElementsWithResource(String name, boolean local
...
+        if (classPathElements.isEmpty()) {
             return ret;
         }
-        ret.addAll(Arrays.asList(classPathElements));
+        ret.addAll(classPathElements);
         return ret;
     }
 
-    public List<String> getLocalClassNames() {
+    public Set<String> getLocalClassNames() {
```
Comment (geoand) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:624]: Why this change of type?

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java [hunk 9]
```diff
@@ -72,9 +72,16 @@ public static boolean isClassPresentAtRuntime(String className) {
...
+            throw new IllegalStateException("The current classloader is not an instance of "
+                    + QuarkusClassLoader.class.getName() + " but " + ccl.getClass().getName());
+        }
+
         String resourceName = fromClassNameToResourceName(className);
-        List<ClassPathElement> res = getElements(resourceName, true);
-        return !res.isEmpty();
+        ClassPathResourceIndex classPathResourceIndex = ((QuarkusClassLoader) ccl).getClassPathResourceIndex();
+
+        return classPathResourceIndex.getFirstClassPathElement(resourceName) != null;
```
Comment (geoand) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java]: This can be simplified a tad by using pattern matching with `instanceof`

**Review by geoand — COMMENTED** [review]
This looks like very work, but I do want some clarification on the how the index works before looking even closer :)

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java (hunk 5 above)
Comment (gsmet) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java:87]: Local resources are resources that are local to your deployment. A jar is an external dependency so it's not local.

This is a preexisting notion as the `QuarkusClassLoader` provides the local resources to the tests: see `QuarkusClassLoader#getLocalClassNames()` in the existing code.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java (hunk 6 above)
Comment (gsmet) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java]: Not really. And interestingly enough, I made some JMH benchmarks this week-end for SmallRye Config (so totally unrelated) and ended up with the version specifying the length being slightly slower.

Note that I conducted some JMH testing of this method to make sure it's fast (not saying it can't be improved but it's not a naive approach to it - see follow-up commit).

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java (hunk 7 above)
Comment (gsmet) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:377]: Sure we could but I don't think it's worse than it used to be. So ideally, let's do that in a follow-up PR?

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java (hunk 8 above)
Comment (gsmet) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:624]: Because that's the semantic we have. A big random bag of unique strings :).

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java (hunk 7 above)
Comment (geoand) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:377]: Sure

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java (hunk 9 above)
Comment (gsmet) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java]: Why not. I don't mind adding some cuteness but I would like to get this patch in and avoid iterating on cosmetic details. I don't think the code is less readable than it used to be.

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java (hunk 5 above)
Comment (geoand) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java:87]: Still sounds confusing to me, but if we are already using that terminology, then 🤷🏽

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java (hunk 8 above)
Comment (geoand) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:624]: Yeah, but what do we gain by changing it to a `Set`?

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java (hunk 9 above)
Comment (geoand) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java]: Sure, I'm just mentioning it because the whole point of the commit was to simplify the method

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java (hunk 2 above)
Comment (gsmet) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java:29]: It does it by storing a prefix instead of the full class name. A prefix that is deemed to be smart enough to not cause a penalty.
As the `ClassPathElement` contains a list of the resources, we can go through the index to reduce greatly the number of elments to go through and then make sure the resource is there.

Comment (gsmet) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java:29]: It's a bit more complex than that as you need to keep full references for some elements (see the sets with local classes, parent first resources...) but that's the general idea of it.

I don't mind adding some javadoc, just tell me if this explanation is good enough for you.

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java (hunk 5 above)
Comment (gsmet) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java:87]: I don't mind changing the name to something that makes more sense to you - I really reused the terminology that was there.
Ideally, let's do that once this patch is in.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java [hunk 10]
```diff
@@ -239,8 +246,8 @@ public Enumeration<URL> getResources(String unsanitisedName, boolean parentAlrea
         //TODO: in theory resources could have been added in dev mode
         //but I don't think this really matters for this code path
         Set<URL> resources = new LinkedHashSet<>();
-        ClassPathElement[] providers = state.loadableResources.get(name);
-        if (providers != null) {
+        List<ClassPathElement> providers = classPathResourceIndex.getClassPathElements(name);
```
Comment (geoand) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:249]: This change will also cause an extra allocation the way we iterate over the list

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java (hunk 8 above)
Comment (gsmet) [reply independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java:624]: The values are coming from Sets now so I decided to be consistent rather than going from a Set to a List, especially since that's the semantic we want.

**Comment by geoand** [comment]
> I personally think it makes things a bit more readable too but YMMV.

Totally agreed

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+
+        public void addBannedResource(ClassPathElement classPathElement, String resource) {
+            bannedResources.add(resource);
+        }
+
+        public ClassPathResourceIndex build() {
+            Map<String, ClassPathElement[]> compactedResourceMapping = new HashMap<>(resourceMapping.size());
+            for (Entry<String, List<ClassPathElement>> resourceMappingEntry : resourceMapping.entrySet()) {
+                compactedResourceMapping.put(resourceMappingEntry.getKey(),
+                        resourceMappingEntry.getValue().toArray(new ClassPathElement[resourceMappingEntry.getValue().size()]));
```
Comment (geoand) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java:234]:
```suggestion
resourceMappingEntry.getValue().toArray(new ClassPathElement[0]));
```

**Review by geoand — APPROVED** [review]
All I really want really at this point is to have the Javadoc of `ClassPathResourceIndex` improved.

Othen than that, all good for me!

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java (hunk 5 above)
Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java:87]: It should be something like `isPathsVisibleOutside()` instead to indicate whether the paths visited while processing this tree are visible outside the context of the `visit` method.

Comment (gsmet) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java:87]: Well, that's not the semantic I want here. Because we could make the jar content visible from the outside and that still wouldn't make the jar resources candidate for this. What I want to really know if we consider these classes local to the current deployment.

BTW, I don't mind people bike shedding on the name but please let's do it in a second step. I really want this in.

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java:87]: > Well, that's not the semantic I want here. Because we could make the jar content visible from the outside and that still wouldn't make the jar resources candidate for this. What I want to really know if we consider these classes local to the current deployment.

From the PathTree perspective that's what this is though. Either the paths are visible or not.

> BTW, I don't mind people bike shedding on the name but please let's do it in a second step. I really want this in.

Sure. That's a public API though, so if it's a follow up it has to go right after this one and if this one gets rolled back, they have to be rolled back together.

**Comment by gsmet** [comment]
@geoand see the additional commits.

The only change compared to what we discussed is the removal of the `name.isEmpty()` workaround. It's no longer needed and I tested that the Liquibase test you introduced when fixing that is passing properly.
(As mentioned in the commit, I fixed it because I had to fix something there anyway)

Now we just have to pray I didn't break anything with these new changes. I didn't some sanity checks but let's see what CI has to say.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java [hunk 12]
```diff
@@ -117,12 +117,12 @@ public List<ClassPathElement> getClassPathElements(String resource) {
...
-        for (ClassPathElement candidate : candidates) {
-            if (candidate.getProvidedResources().contains(resource)) {
-                classPathElements.add(candidate);
+        for (int i = 0; i < candidates.length; i++) {
+            if (candidates[i].getProvidedResources().contains(resource)) {
+                classPathElements.add(candidates[i]);
             }
         }
-        return Collections.unmodifiableList(classPathElements);
+        return classPathElements;
```
Comment (gsmet) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathResourceIndex.java:141]: I think given it's a hot path and not really user-consumable APIs, we could get rid of this safety.

**Comment by geoand** [comment]
PR looks good to me.

The additional commits seem like overkill to me, but I don't care enough about that kind of thing to fight it :)

**Comment by gsmet** [comment]
@geoand didn’t you ask about these changes this morning? (At least the for loops and the pattern matching)

**Comment by geoand** [comment]
Right, what I meant is that having those changes as additional commits seems like overkill to me - my bad for not clarifying

**Comment by gsmet** [comment]
Oh I wanted to make it easy for you to review them. I had plans to squash some of the commits (if not too much of a mess). Also I wanted a full CI run green before that to make sure I didn’t break my patch.

**Comment by geoand** [comment]
🙏

**Review by geoand — APPROVED** [review]

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java (hunk 5 above)
Comment (gsmet) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java:87]: @aloubyansky I don't think you should place yourself from the PathTree perspective when you define the API: it's really for the consumers that the name is important.

For this very patch, I don't want to know if I'm able to get the content using something else than visit. I want to know if this thing is an archive and thus not something that will be subject to change.
The fact that I have to use `visit` is really an implementation detail (and could change over time) and moreover, that's not what I'm trying to do here: I'm just trying to know if the classes will be subject to change or not.

I'm not saying that my name is ideal but at least from my perspective `isPathsVisibleOutside` has an obscure semantic and I have no idea what I should use it for. And I probably won't have used it if it existed as for me it's not the same semantic.

If you come with a better name, I'll be happy to follow up after this PR is in.

Comment (geoand) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java:87]: > thus not something that will be subject to change.

So why not come up with a name that reflects that?

Comment (gsmet) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java:87]: I'm all ears but really it can be done in a follow-up as this patch have been around for quite a while now and I want it in.
If you all can come up with a good proposal, I will be happy to do the leg work in another PR.

**Comment by gsmet** [comment]
I squashed the commits that were just tiny improvements on top of the rest and kept the ones with a specific semantic that I want to keep around.

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java (hunk 5 above)
Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java:87]: > isPathsVisibleOutside has an obscure semantic

Local is a term from a higher (classloading) level. It doesn't apply well to `PathTree` abstraction. `isPathsValidOutside()` is exactly what this is.

Comment (aloubyansky) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java:87]: @gsmet is your concern about `OpenContainerPathTree`? Would `isArchiveOrigin()` be better for you?

Comment (gsmet) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java]: Well, I don't think it is, especially because of the semantic in `MultiRootPathTree` but I pushed the change as I'm a bit tired of arguing.
My personal opinion is that the semantic I chose was good and maybe the name needed some love. Maybe `providesApplicationClasses` to be on par with `QuarkusClassLoader#isApplicationClass()` as it's really the same semantic.
But I don't think the name was that bad once you understand the semantic.

I spent a lot of time on this patch and I would really appreciate if we could take a decision and merge it.

Thanks.

Comment (gsmet) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/PathTree.java]: Note that if you don't understand the semantic, we might better have a quick call so that I explain it to you.

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathElement.java [hunk 13]
```diff
@@ -74,9 +74,11 @@ default ResolvedDependency getResolvedDependency() {
     Set<String> getProvidedResources();
 
     /**
-     * Whether this class path element contains local resources.
+     * Whether the content of this tree comes from an archive or not.
+     * <p>
+     * This is useful for instance when you want to determine if the resources can be updated in dev mode.
      */
-    boolean providesLocalResources();
```
Comment (aloubyansky) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/ClassPathElement.java]: My comments were about `PathTree` API only. This is a layer on top of it where `PathTree` is an impl detail and so here the methods could reflect the meaning that's relevant in this context.

**Comment by gsmet** [comment]
@aloubyansky I rewrote the last commit as per our discussion. I think we should be all good now but I would appreciate some extra scrutiny. Thanks!

**Review by aloubyansky — APPROVED** [review]
